cla_backend/apps/knowledgebase/tests/test_csv_2_fixture.py

```diff
@@ -0,0 +1,1443 @@
+from django.utils.dateparse import parse_datetime
+
+from knowledgebase.management.commands._csv_2_fixture import KnowledgebaseCsvParse
+
+
+class TestCSV2Fixture(TestCase):
+    def setUp(self):
+        self.datetime_now = datetime.now().replace(tzinfo=timezone.get_current_timezone()).isoformat()
+
+    def calculate_pk_range(self, article_category_matrices):
+        min_range = min(article_category_matrices, key=lambda x: x["pk"])
+        max_range = max(article_category_matrices, key=lambda x: x["pk"])
+        return min_range["pk"], max_range["pk"]
+
+    def sort_article_category_matrices(self, output_category_matrix, expected_category_matrix):
+        output_sorted_acm = sorted(output_category_matrix, key=lambda x: x["fields"]["article_category"])
+        expected_sorted_acm = sorted(expected_category_matrix, key=lambda x: x["fields"]["article_category"])
+        return output_sorted_acm, expected_sorted_acm
+
+    def slice_records(self, output_records, expected_records, startRange=None, endRange=None):
+        output_sliced = output_records[startRange:endRange]
+        expected_sliced = expected_records[startRange:endRange]
+        return output_sliced, expected_sliced
+
+    def compare_dicts(self, outputDict, expectedDict):
+        for key, value in expectedDict.items():
+            self.assertIsInstance(value, type(outputDict[key]))
+            if isinstance(value, int):
+                self.assertEqual(outputDict[key], value)
+            elif isinstance(parse_datetime(value), datetime):
+                self.compare_datetimes(parse_datetime(outputDict[key]), parse_datetime(value))
+            else:
+                self.assertEqual(outputDict[key], value)
+
+    def compare_datetimes(self, outputDatetime, expectedDatetime):
+        time_diff = outputDatetime - expectedDatetime
+        time_diff_in_seconds = time_diff.total_seconds()
+        self.assertLessEqual(time_diff_in_seconds, 0.5)
+
+    def compare_list_of_records(self, outputList, expectedList):
+        for expectedObject, outputObject in zip(expectedList, outputList):
+            self.compare_length(outputObject, expectedObject)
+            for key, value in expectedObject.items():
+                self.assertIsInstance(value, type(outputObject[key]))
+                if isinstance(value, dict):
+                    self.compare_dicts(outputObject[key], value)
+                else:
+                    self.assertEqual(outputObject[key], value)
+
+    def compare_length(self, output, expected):
+        self.assertAlmostEqual(len(output), len(expected))
+
+    def compare_article_category_matrices(self, output, expected, pk_range):
+        for outputObj, expectedObj in zip(output, expected):
+            self.compare_length(output, expected)
+            for key, value in expectedObj.items():
+                if key == "pk":
+                    self.assertIsInstance(value, int)
+                    self.assertIsInstance(value, type(outputObj[key]))
```

Comment:
What does this check do? It looks to me like it checks the same thing as the line before but in a more generic way (ie it'd work even if we weren't using ints for pks, but then the line before would fail?)

Actually, I've read the other checks and I think it's just checking that the output and expected values are of the same type? This is confusing for a couple of reasons.
1) The check is that the expected value is the same as the output one. It's not a big thing, but this being back-to-front probably helped confuse me
2) It doesn't actually check the types are the same. It checks the first is the same or some subclass of the second. If you do want to check they're the same, use `assertEqual` instead, with `type()` of both values

Reply:
Actually line 63 doesn't check the values are the same. It just checks the expected value is an integer. That doesn't do anything by itself, so by adding the next check means I can ensure my output from new functionality is an integer.

My issue was we know we want the value to specifically be an integer, hence line 63, but I needed a way to compare that, hence line 64.

Also done it like this because we don't know the actual value.

Reply:
I didn't suggest to compare the values here

Reply:
Ah okay my bad

Reply:
We can talk about this on our next pairing